GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs

```diff
@@ -0,0 +1,89 @@
+﻿using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+
+namespace GVFS.Common.Maintenance
+{
+    public class LooseObjectsStep : GitMaintenanceStep
+    {
+        public const string LooseObjectsFileName = "loose-objects.time";
+        private readonly bool forceRun;
+
+        public LooseObjectsStep(GVFSContext context, bool requireCacheLock, bool forceRun = false)
+            : base(context, gitObjects: null, requireObjectCacheLock: requireCacheLock)
+        {
+            this.TimeFilePath = Path.Combine(this.Context.Enlistment.GitObjectsRoot, "info", LooseObjectsFileName);
+            this.forceRun = forceRun;
+        }
+
+        public override string Area => nameof(LooseObjectsStep);
+
+        protected override TimeSpan TimeBetweenRuns => TimeSpan.FromDays(7);
+
+        public int CountLooseObjects()
+        {
+            int count = 0;
+            
+            foreach (string directoryPath in this.Context.FileSystem.EnumerateDirectories(this.Context.Enlistment.GitObjectsRoot))
+            {
+                string directoryName = directoryPath.TrimEnd(Path.DirectorySeparatorChar).Split(Path.DirectorySeparatorChar).Last();
+
+                if (SHA1Util.IsValidHex(directoryName, 2))
+                {                    
+                    count += this.Context.FileSystem.GetFiles(Path.Combine(this.Context.Enlistment.GitObjectsRoot, directoryPath), "*").Count();
+                }
+            }
+
+            return count;
+        }
+
+        protected override void PerformMaintenance()
+        {
+            using (ITracer activity = this.Context.Tracer.StartActivity(this.Area, EventLevel.Informational, Keywords.Telemetry, metadata: null))
+            {
+                try
+                {
+                    if (!this.forceRun)
+                    {
+                        if (!this.EnoughTimeBetweenRuns())
+                        {
+                            activity.RelatedWarning("Skipping step due to not enough time between runs");
+                            return;
+                        }
+
+                        // Verify no git processes are running
+                        IEnumerable<int> processIds = this.RunningGitProcessIds();
+                        if (processIds.Any())

```

Comment:
Would we still want to check for git process even in a forced run?  What could happen if there is a git process running that would affect the prune-packed?

Reply:
@derrickstolee and I talked about this offline and the forceRun is only set for functional tests.  We want to force certain scenarios here so yes we'd like it to run even if a git command is going.  We will add more testing around particular git commands.

Reply:
I should certainly add a comment that forceRun is only done in testing for clarity.  I'll do that!